### DISCUSSION TITLE
Seeking a way to improve the documentation, when, how, who?

### DISCUSSION THREAD

**Discussion by ksaadDE** [diesel-rs/diesel#4099] [Q&A]
As show of good will, I went through the past Q&As and made a list of common asked questions. So please take a look at it and understand the pattern, which is not coming from the community. It is clearly a guide issue!

> How to fix? It doesn't work for me: cannot find function insert_into in this scope […]

User was told:
> The error message already points out the correct solution

So user replied:
> Which made me question **if the guide is wrong, or was not updated**. So here I am. [...] Why write a guide […] when at its end you refer to the sample project anyway? 

User was told:
> I personally do not plan to change the guide just because **one person [actually multiple people complained!]** complained about this

---

https://github.com/diesel-rs/diesel/discussions/4089

> I cannot find detailed chapters for SELECT. Any pointers? I'm looking for ways of doing SELECT with multiple filters (i.e. filter on multiple columns). […] how do we use filter methods to implement SQL IN operator?  […] I think I found it. eq_any() seems to be.

User was told "currently no complete guide for select statements" and that all methods for that  are document in trait.QueryDsl within the docs. 

#2735 https://github.com/diesel-rs/diesel/discussions/2735#discussion-3325997

---

> I've been having troubling finding a good and ergonomic way to implement these various traits for my custom types […] A best-practices guide for implementing the Diesel ORM traits would be a great and much sought-after addition to 

User was told "Putting everything behind one derive is just not an option as there are a lot of cases where only a subset of the conversions offered by the different derives are valid or even possible"

#2938 https://github.com/diesel-rs/diesel/discussions/2938#discussion-3659582

---

> it seems that I had some confusion about migrations […] The following documentation is misleading, mod embedded_migrations is actually created by the embed_migrations macro, it would be helpful if the documentation actually stated that to avoid confusion […] Based on the example, I thought I would need to add a use statement that included both the embed_migrations! macro as well as the embedded_migrations module.

#3030 https://github.com/diesel-rs/diesel/discussions/3030#discussioncomment-2054532  

---

> I have read this guide, but I'm not really sure if I understood correctly […] I have 4 lines of typing. I mean, is there a way to make this more ergonomic? […] It's just that writting the return type felt like rewriting the query itself […] it felt kinda wrong.

User was told "That's currently the only way to specify these return types"

#3458 https://github.com/diesel-rs/diesel/discussions/3458#discussion-4678292

---

Hard time to generate a equivalent of a SQL statement 

#3513 https://github.com/diesel-rs/diesel/discussions/3513#discussion-4859618

---

User can't find the best way to update many multiple records with a complex scenario including AsChangeSet and objects using it (not covered in the guide; to find batch queries are not implemented)

https://github.com/diesel-rs/diesel/discussions/3417#discussioncomment-4450367

---

> The pagination example from our guide is specifically not an example how to implement a optimized paginate query. It's an example how to extend diesels query builder, that just happens to use an fitting query.

#3096 https://github.com/diesel-rs/diesel/discussions/3096

---

> guide seems a little outdated

#2890 https://github.com/diesel-rs/diesel/discussions/2890#discussioncomment-1280961

---

custom schemas were not findable in the guide

#2990 https://github.com/diesel-rs/diesel/discussions/2990

---

> tried to see some examples on the official website page, which I found difficult to understand

https://github.com/diesel-rs/diesel/discussions/3162

---

> his snippet it requires too much work to know where establish_connection() comes from for example.

https://github.com/diesel-rs/diesel/discussions/3162#discussioncomment-3758788

(with you commenting to another post: "Please consider to take more time to **formulate your questions in such a way that they are understandable by other people [unlike the guide xD]**.")

---

User was unable to find "Extending Diesel", because user did not know whether "lateral queries" were part of diesel

https://github.com/diesel-rs/diesel/discussions/3132#discussion-4010220

---

> I am soliciting your help after several unsuccessful attempts [...] new diesel type […]. special feature […] AutoIncrementID(Option<i64>); 
The user then got told that ToSQL is wrong for that, because SQL queries can not be run in it, instead SQL default values should be used. A link to the relevant guide was placed in the reply.
Core issue, user did not know that ToSQL cannot be used for it and that SQL default values are a better option for it

https://github.com/diesel-rs/diesel/discussions/4031

---

Users are often recommended to not implement generic code, as it gets hard to maintain (which probably conflicts with most basic programming best-practices out there). Users should also look up the type defs in the docs for the corresponding methods.

https://github.com/diesel-rs/diesel/discussions/3968#discussioncomment-8844810

---

> The docs didn't really help me out with this, as it's really hard to find it […] I could try to use IIF statements, if they are available [docs do not state if they are available]  […] there is no working example that shows how to actually use or implement it, which is exactly what you are complaining about (and why I actually put that aside)"

User was told:
- > "For the "missing" case expression: As already written you can use the master branch"
- > "For IFF you need to roll your own implementation"
- > and "That's in implementation detail, that is not relevant for using that macro"

Nice to know, I wish the guide/docs would somehow mention that. 
https://github.com/diesel-rs/diesel/discussions/3947

---

> Something is wrong with the offical guide
https://github.com/diesel-rs/diesel/issues/1551

---

It's just a example, minimal set, obtained from searching for all Q&As containing "guide"! Are all of us "stupid idiots" (see #4089) that can not read your guide nor the RUST documentation properly? 🤔

**Comment by weiznich** [comment]
To repeat again what I already have written before: I do own you nor anybody else nothing. Please stop writing these posts, they are not helpful.

It's generally really bad style to accuse maintainers of not doing their work and requesting them to do a certain thing. If you really need to have something done consider funding the feature/documentation/… you want to have. Otherwise I do not have any obligation to do anything for you. (And given you attitude I won't do anything for you).

That written: We are aware that more documentation would be great to have (as for ~any other project out there), but in the end someone needs to write that documentation. Given that other people request more features or fixing certain bugs or something else there is just not enough time to work on everything at the same time. Also I do have my own use cases, that I prioritize for obvious reasons. The hard truth there is documentation does not pay me anything, while adding/polishing new features for products I develop does pay me something. In the end it boils down to: If you want to have better documentation (or anything else) the best way to make sure that this happens is to start contributing. Otherwise you need to wait until someone has time, skill and motivation to fix the issue. Instead you chose to harass the maintainer, which is a really bad way to make things happen. (Hint: As soon as a person tries to press me in doing something or just claims that this obviously is broken and needs to be fixed by me now I likely stop doing anything there for self protection for some time).

Given these facts it's just not helpful to have you repeated posts complaining about documentation and guides here. If you don't stop with that I block you from the organization as well, as this just takes more time away from work that could be done otherwise.

**Reply by ksaadDE** [reply]
# Your Acknowledgement

> We are aware that more documentation would be great to have

Thank you! I am glad you finally acknowledge that significant issue.  Late is better than never.  

All I am asking you for the next times, instead of insulting or degrading our knowledge level, please consider our needs as well without ignoring your own ones.

I really appreciate your absolutely understandable explanation and your call for assistance (contributions)!  I hope we can tackle it - exactly what open source and community projects are about.

The beauty of (F)OSS is we can work together and increase our common profit - which in return increases yours.

I never pressured you into something. If it feels like pressure, it's soley on your side and indicates we should tackle it.  I am pointing out the core issue, which is valid critique and I think blocking me for sound and valid, voiced critique would be a pretty weak response. Better it's me pointing the issue out than others. Think about that perspective. 

# Lacking documentation and Revenue

> The hard truth there is **documentation does not pay me anything**, while adding/polishing new features for products I develop does pay me something. 

While I am absolutely fine with your other responsibilities and what ever floats your boat, lacking documentation costs you a lot more (compared to what you would possibly earn) in the end. 

## Insufficient Documentation Lowers Your Reveneue
1. Users ask questions that you would otherwise have not being asked or Users ask multiple of the same typed questions (as the core difficulties of the docs are not resolved), both requires additional time overhead, stealing your product improvement time
2. Compliance issues over time (for your customers and yourself), even with a "given as is" license, such legal fees are much higher than actually fixing the docs
3. Lower satisfaction of the community leads to less contribution, harms the project overall and might scare your customers away 

**Example one:**
> The cost of developing a typical end user document at Cadence Design Systems (a 300-page reference or user manual or the online equivalent) is about $40,000. The cost of not providing complete and completely accurate documentation can be in the hundreds of thousands of dollars. Investing more pre-sales dollars in timely, accurate, and complete documentation may significantly **reduce the total cost  [especially time] of customer support.** https://www.ingentaconnect.com/content/stc/tc/1995/00000042/00000001/art00009

**Example two:**
> Suppose paying intermediaries to clean up data and **improve documentation quality costs less** than data **producers to do the work themselves**, Intermediaries, such as data archives, become a lower cost solution for data producers
https://asistdl.onlinelibrary.wiley.com/doi/full/10.1002/meet.2009.145046024

## Providing sufficient documentation is a well known industry standard, even in Open Source
1. https://www.linkedin.com/pulse/common-documentation-problems-how-solve-them-douglas-black
5. https://doakio.com/blog/the-hidden-cost-of-poor-documentation-a-quantitative-study/
6. https://dataedo.com/blog/cost-of-no-data-documentation

## What's actually needed?
I believe:
- We need a short guide that is understandable and easy for beginners, similar to the [existing guides](https://diesel.rs/guides/); smoothening the steep learning curve - The guides are technically well explaining (I assume) but they lack a connection to the user's pov - why are we doing xyz? What are we trying to achieve? They lack that kind of common thread
- Coverage of really basic use-cases like the example project but a bit more extended for common use e.g. with web servers + crud, pagination, statistics, rudimentary user-password login-systems, API stuff etc, so we users can adapt\*
- A technical documentation for the pros (where all the little API pieces and their relations are looked at)
- Some kind of Quality Management and Feedback-Improvement Loop (for continuously improving and creating the overview)

\*Possibly if you want to stick with method-based [examples](https://github.com/diesel-rs/diesel/tree/master/examples), we need a more narrow focus on CRUD. What I see is a confusing mix of technical concepts (the Rust lingo, the Diesel lingo) and CRUD.  The code examples in the Repo making it also difficult to understand what exactly I am looking at and what I need that for. 

## Documentation Size and Extend
Nobody wants a perfectly written 300 page document. That's not necessary and would require too much time investment and work. So strike that.

Most people, including me, prefer a basic  e.g. what we are going to do/solve (introduction), common use-cases, common issues, and how to get things running asap. 

Allowing us to copy paste basic things and then work from there. So our creativity and fun flows. It allows us to build a work routine, enabling us to help others - consequently reducing your time efforts in the Q&A section.

I have often seen a three-level guide, from beginner, intermediate to pro. Not sure if we need that but that could be an idea to satisfy the needs of all.

## Consider how the general public reads and uses documentations

I think we should consider how the general public reads such documentations and guides. Additional information and better documentation increases the likelihood of good contributions.  

Nobody reads it all, people skim over text, search for easier explanations and simple tutorials etc. People learn in small steps over time. Too much info and too much effort to get into something literally throws them off and prevents them from contributing.  

**Examples for online tutorials**
- https://blog.logrocket.com/interacting-databases-rust-diesel-vs-sqlx/
- https://www.youtube.com/watch?v=tRC4EIKhMzw
- https://medium.com/@rameshovyas/a-step-by-step-guide-to-access-mysql-database-in-rust-using-diesel-orm-41836062bb6d
- https://planetscale.com/blog/build-a-rust-api-with-rocket-diesel-mysql
- https://dev.to/turboza/using-datetime-in-rust-diesel-with-mysql-31ib
- https://cprimozic.net/blog/rust-rocket-cloud-run/
- https://duredhelfinceleb.github.io/diesel_windows_install_doc/
- https://kitsu.me/posts/2020_05_24_custom_types_in_diesel
- https://qiita.com/Gma_Gama/items/a489be2207f0b35f9282
- https://ozway.jp/2020/10/rust-mysql-diesel%EF%BC%881%EF%BC%89/
- https://dev.to/ssivakumar/rust-diesel-fixing-issues-with-setup-3k56
- https://yujinping.top/post/rust/build_web_api_with_rust_rocket_-diesel_mysql/
- https://www.youtube.com/watch?v=_pPtsPNdNrc
- https://fdeantoni.medium.com/rust-actix-diesel-sqlite-d67a1c3ef0e
- http://profectium.blogspot.com/2017/11/rust-setting-up-sqlite-database-with.html

# What prevents a good Quality Management (QM)?

**Three root-causes preventing contributions in the documents' Quality Management:**
- No overview of what needs more documentation (or less) 
- Lack of insight in the internals hindering contributions  (Without deep understanding, documenting it better gets difficult)
- Negative atmosphere and attitude towards the community (especially when trivial questions are asked)
- Maybe a missing one place overview about what has been often asked or already answered (compare it to e.g. https://www.reddit.com/r/rust/comments/119we2q/new_diesel_guide/ and the repo issues)

# Contributions

## Defining contributions
Do we have the same understanding what contributing is? If you ask me to increase your business output, that is solely your responsibility, not mine. I already provided actionable feedback, collected or analyzed multiple times where more improvements are needed. I even contributed to it's fixing.

## Contributions relations to atmosphere and insight (knowledge) level

You want contributions but at the same time we are getting insulted, degraded or told off, when asking questions, voicing critique or just complaints and contributing.

Also, insufficient documentation reduces the insight of the community, consequently reducing the ability to improve the documentation. Like previously shown basic areas are affected, especially the core understanding how to tackle specific issues on our own, the same way we would need a instruction set on how to debug a software of any software company. So a general scheme what should we do first to possibly come to a working solution, maybe even before asking a question.

Mentioned points decreasing the likelihood of receiving overall contributions, not just documentation QM related ones. Looks a bit like a chicken-egg problem to me, a bit difficult to tackle but possible.

### How I already contributed

Regarding 
> make sure that this happens is to start contributing

### I helped establishing a basic overview and some QM

I did a first step in actually documenting such issues and highlighting areas where the documentation or guides have weak spots. I went through the Q&A issues which explicitly mention the guide (see above). Gladly, we can appreciate their long-term existence, allowing that kind of collection and analysis of them. 

### I worked on write ups and asked questions

Additionally, I already wrote a small write up regarding the enum and sqlite issue (see #3742 and https://github.com/ksaadDE/rustgotchas/blob/main/How_to_convert_a_Enum_to_DB_row_string_value_using_Diesel.md)

**Once provided, you've bashed against the contribution:**
> Please be aware that this document only covers the specific use of that single user, mixes in quite a bit of unrelated stuff, and misses some quite important information

Sure, it might not be perfect but it would have been helpful once getting into Diesel. Good enough to get the boat floating. Details are a matter of detail and specifics. Expertise develops over time, while making progress.

I would have written another one but sadly my question https://github.com/diesel-rs/diesel/discussions/3947 was locked, after I was insulted (yet again). It was about implementing a enum-like if condition for multiple columns of a db row on db-level to use the calculated result later on (e.g. result = row.a && row.a && !row.c => 1; result = row.a && !row.b && row.c => 2 etc).

Yet again I spent some time to contribute and help out. Like I did with many other projects in my limited spare time.

**Reply by ksaadDE** [reply]
Little addendum of critique or lacking understandings by others (those are from the web and likely a bit older) :
>  […] Diesel is a lovely query builder but whenever I try to use it for something other than postgres **I always run into friction.** 
https://www.reddit.com/r/rust/comments/dtz49q/typescript_in_electron_rust_diesel_sqlite/

> Why does everyone recommend diesel when **it's so confusing to use**? [...] Most of the tutorials I found use diesel to perform the schema translation. However, **after about 3 hours of wrestling with diesel I just gave up** and decided to switch to vanilla rust::postgres and manage my schema and migrations separately. 
https://www.reddit.com/r/rust/comments/g07t7a/why_does_everyone_recommend_diesel_when_its_so/

> You need to enable sqlite as a feature in your Cargo.toml
https://www.reddit.com/r/rust/comments/4puixn/how_do_you_use_sqlite_with_diesel/

Someone had issues finding the guide to MyQL 
https://www.reddit.com/r/rust/comments/f0drzn/connecting_mysql_database_with_diesel_if_possible/

Connecting to MySQL was a problem in the past too, seemingly the user had trouble to disable the SSL mode to make the connection
> documentation says default for ssl_mode is PREFERRED. If we can change it to DISABLED, that might be another solution.
https://users.rust-lang.org/t/unable-to-connect-to-mysql-using-diesel/93949/3

Seemingly the starting guide also caused trouble (due to minimal changes over time? due to misreadings?)
- https://stackoverflow.com/questions/75398741/problems-following-diesel-getting-started-guide-with-sqlite
- https://stackoverflow.com/questions/67815883/following-getting-started-diesel-tutorial-but-with-sqlite-causes

User says they need to get used to the expression stuff
https://stackoverflow.com/questions/76092639/following-tutorial-for-rust-diesel-getting-stuck-with-dieselqueryable-not-sa

Diesel and Windows - ure referring to a bundle in there. First time I have seen this :D
https://users.rust-lang.org/t/struggling-with-diesel-sqlite-on-windows/73336/4

Another one has a diesel setup issue but fixed it using a `pg_hba.conf`
> Ok, I've managed to connect to a database. The issue was pg_hba.conf file with ident instead of md5.
https://users.rust-lang.org/t/diesel-setup-fails/110006/4

I had that same question at the beginning  "how-to-make-diesel-auto-generate-models"
https://stackoverflow.com/questions/70410103/how-to-make-diesel-auto-generate-models

Unlucky user errors in handling Rust (or better to say their env varibales of the OS) are also happening:
https://www.reddit.com/r/rust/comments/1cbm0yy/diesel_problems/

>  I have a MySQL database and I want to make the simplest possible query in rust to verify that I'm connecting to the database and able to retrieve data
https://stackoverflow.com/questions/77897292/generic-type-of-raw-sql-query-in-rust

> all the inserts are completely synchronous. How can I make it so that it doesn't require as much back and forth from the database? Or at least make all the request be sent at once?
https://stackoverflow.com/questions/77922820/how-to-bulk-inserts-returning-queries-with-diesel-rocket-and-postgresql

User needed five hours to find a wrong order of is_admin and password (error message is about derserialize and uuid)
https://stackoverflow.com/questions/77298331/diesel-2-fails-to-deserialize-uuid

Excellent explanation: "Why does order of fields in struct need to match table? TL;DR: Diesel prefers deserializing by index instead of by name." (even with ref. to the docs!)
https://stackoverflow.com/a/77169379

headache with delete return val
https://stackoverflow.com/questions/77150457/return-type-of-row-deletion-function-in-rust-using-diesel

How to use upsert + insert_into + on conflict +  excluded +  AsChangeset
https://stackoverflow.com/a/47626103 

> You are also missing the password_hash field that your table has. Either add it to the model definition, or if you have omitted it on purpose you can add a SELECT clause to only get those fields (since you've already derived Selectable):
https://stackoverflow.com/a/76750613

Mapping a postgres view?
https://stackoverflow.com/questions/76749491/mapping-postgres-view-in-diesel-rust-issues

> Your conflict logic doesn't do anything in the database so there's no reason for Diesel to support it. Instead of telling your database to silently ignore conflicts, just do the insert and let it return an error. You would get back a [Error::DatabaseError](https://docs.diesel.rs/master/diesel/result/enum.Error.html) with [DatabaseErrorKind::UniqueViolation](https://docs.diesel.rs/master/diesel/result/enum.DatabaseErrorKind.html) if there was a conflict and you can detect that and do your own logic afterwards.
https://stackoverflow.com/a/76694571

> You need to enable the "chrono" feature for the implementation for DateTime<UTC> from the  
https://stackoverflow.com/a/76384410 (btw I had that issue too, the solution did not work for me, if I remember right)

> Select clauses cannot mix aggregated and non aggregated expressions. [... as] they can return non-deterministic results [...]
https://stackoverflow.com/a/76012050

How to insert a decimal number with Diesel's PgNumeric type?
https://stackoverflow.com/a/38175685

Not sure if still valid: [Creating Diesel.rs queries with a dynamic number of .and()'s](https://stackoverflow.com/questions/48696290/creating-diesel-rs-queries-with-a-dynamic-number-of-ands)

[How to display the value retrieved using Diesel in HTML using a Tera template in Rocket?](https://stackoverflow.com/questions/42063537/how-to-display-the-value-retrieved-using-diesel-in-html-using-a-tera-template-in)


As far as I've seen many Questions, misunderstanding and problems can also be extracted from https://stackoverflow.com/questions/tagged/rust-diesel

**Reply by weiznich** [reply]
As you already done all the work to put together this list of things that you personally consider badly documented I would suggest that you contribute to fix these issues. After all that seems to be something that you have feelings about. I personally do not agree about quite a large number of these points, but that's a personal opinion and nothing I would reject better documentation for. Additionally almost all of your links refer to old diesel versions so it's easy to claim that these are still an issue. Would you have spend the time to actual check that, you would have noticed that some significant part of this is already addressed by various "new" features and documentation changes.

As written before: Due to other priorities I do not have the capacity for that and I do also not have the motivation to continue this toxic discussion. Consider this my last answer to any of your requests until you fundamentally change your wording and your approach how you interact with this project. You need to understand that there is a fundamental difference between a user of this project and the maintainers, especially if you ask for help. The most significant point there is that you want something from me not the other way around. I do not force you to use my project, nor I'm interested in that you personally use it, while you want to know how to solve your issue from me.

**Reply by ksaadDE** [reply]
I kindly ask to apologize the miscommunication about my previous questions and suggestions.

Generally I only raise questions after spending hours searching for information, ending up frustrated, unable to proceed and then I am only seeking to speed things up. Your maintainer insights and wisdom would have been helpful to enable me to solve my issues on my own, avoiding future questions by telling/showing us how you think in Diesel terms. It would have helped me to convey that knowledge to others.

I understand you need more contributions and you prioritize eminent financial aspects to make Diesel more awesome. Reoccurring/"stupid questions" and "lacking contribution" (= wasted time => less money for you) indicating that a few time investments in sufficient documentation are needed. 

My list might not be sufficient as a good starting point. It lacks the maintainers perspective, organization, structure and internal knowledge. Latter is a reason why we users are yet unable to contribute.

Knowledge gaps that I see:
- What areas have reoccurring issues and questions been about? 
- What needs to be understood when beginning using Diesel?
- What is important when implementing common use-cases in Diesel? (e.g. common pitfalls) 
- Are there technical specialties in Diesel that we need to consider? 

Where do we start? What list/overview?

Filling the knowledge gap will likely make your project more public and attract more contributors.
